python_scripts/dev_tree.py

```diff
@@ -0,0 +1,783 @@
+ax.plot(
+    X_test_subset, y_pred, label="Test predictions",
+    color="tab:green", marker="^", markersize=10, linestyle="",
+)
+
+plt.legend()
+# %% [markdown]
+# In the contrary of the linear model, the decision trees are non-parametric
+# models and do not rely on the way data should be distributed. In this
+# regard, it will affect the way the model will predict. We can repeat the
+# above experiment and see the differences.
+
+# %%
+from sklearn.tree import DecisionTreeRegressor
+
+tree = DecisionTreeRegressor()
+
+# %%
+_ = plot_regression_model(X_train, y_train, tree)
+
+# %% [markdown]
+# We see that the decision tree model does not have an a-priori and we do not
+# end-up with a straight line to regress flipper length and body mass. We can
+# observe that when we generated a sample which was as well present in the
+# training set, the tree will output the same target than this training sample.
+# However, if for the same flipper length, we have different body masses, then
+# the tree is predicting the mean of the targets.
+#
+# So in classification setting, we saw that the predicted value was the most
+# probable value in the node of the tree. In the case of regression, the
+# predicted value corresponds to the mean of the target in the node.
+#
+# This lead us to question whether or not our decision trees are able to
+# extrapolate to unseen data. We can highlight that this is possible with the
+# linear model because it is a parametric model.
+
+# %%
+plot_regression_model(X_train, y_train, linear_model, extrapolate=True)
+
+# %% [markdown]
+# The linear model will extrapolate using the fitted model for flipper length
+# < 175 mm and > 235 mm. Let's the difference with the trees.
+
+# %%
+ax = plot_regression_model(X_train, y_train, linear_model, extrapolate=True)
+_ = plot_regression_model(X_train, y_train, tree, extrapolate=True, ax=ax)
+
+# %% [markdown]
+# For the tree, we see that it cannot extrapolate below and above the minimum
+# and maximum, respectively, of the flipper length encountered during the
+# training. Indeed, we are predicting the minimum and maximum values of the
+# training set.
+#
+# ### The regression criterion
+#
+# In the previous section, we explained the differences between using decision
+# tree in classification or in regression: the predicted value will be the
+# most probable class for the classification case while the it will be the mean
+# in the case of the regression. The second difference that we already
+# mentioned is the criterion. The classification criterion cannot be applied
+# in regression setting and we need to use a specific set of criterion.
+#
+# One of the criterion that can be used in regression is the mean squared
+# error. In this case, we will compute this criterion in each of the partition
+# as in the case of the entropy and select the split leading to the best
+# improvement (i.e. information gain).
+#
+# ## Importance of decision tree parameters on generalization
```

Comment:
```suggestion
# ## Importance of decision tree hyper-parameters on generalization
```

Reply:
I am not sure if we shall use "hyper parameters" rather than "parameters" here and below.